be/src/runtime/cache/result_node.cpp

```diff
@@ -0,0 +1,274 @@
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+#include "gen_cpp/internal_service.pb.h"
+#include "runtime/cache/result_node.h"
+#include "runtime/cache/cache_utils.h"
+
+namespace doris {
+
+bool compare_partition(const PartitionRowBatch* left_node, const PartitionRowBatch* right_node) {
+    return left_node->get_partition_key() < right_node->get_partition_key();
+}
+
+//return new batch size,only include the size of PRowBatch
+void PartitionRowBatch::set_row_batch(const PCacheValue& value) {
+    if (_cache_value != NULL && !check_newer(value.param())) {
+        LOG(WARNING) << "set old version data, cache ver:" << _cache_value->param().last_version()
+                     << ",cache time:" << _cache_value->param().last_version_time()
+                     << ", setdata ver:" << value.param().last_version()
+                     << ",setdata time:" << value.param().last_version_time();
+        return;
+    }
+    SAFE_DELETE(_cache_value);
+    _cache_value = new PCacheValue(value);
+    _data_size += _cache_value->data_size();
+    _cache_stat.update();
+    LOG(INFO) << "finish set row batch, row num:" << _cache_value->row_size()
+              << ", data size:" << _data_size;
+}
+
+bool PartitionRowBatch::is_hit_cache(const PCacheParam& param) {
+    if (param.partition_key() != _partition_key) {
+        return false;
+    }
+    if (!check_match(param)) {
+        return false;
+    }
+    _cache_stat.query();
+    return true;
+}
+
+void PartitionRowBatch::clear() {
+    LOG(INFO) << "clear partition rowbatch.";
+    SAFE_DELETE(_cache_value);
+    _partition_key = 0;
+    _data_size = 0;
+    _cache_stat.init();
+}
+
+/**
+ * Update partition cache data, find RowBatch from partition map by partition key,
+ * the partition rowbatch are stored in the order of partition keys
+ */
+PCacheStatus ResultNode::update_partition(const PUpdateCacheRequest* request, bool& update_first) {
+    update_first = false;
+    if (_sql_key != request->sql_key()) {
+        LOG(INFO) << "no match sql_key " << request->sql_key().hi() << request->sql_key().lo();
+        return PCacheStatus::PARAM_ERROR;
+    }
+
+    if (request->value_size() > config::query_cache_max_partition_count) {
+        LOG(WARNING) << "too many partitions size:" << request->value_size();
+        return PCacheStatus::PARAM_ERROR;
+    }
+
+    //Only one thread per SQL key can update the cache
+    CacheWriteLock write_lock(_node_mtx);
+
+    int64 first_key = kint64max;
+    if (_partition_list.size() == 0) {
+        update_first = true;
```

Comment:
'update_first' 's meaning is not clear.  Maybe you can call it  'is_first_key_updated'